File: examples/basic_usage.py
"""Basic usage example for Jules Agent SDK."""

import os
from jules_agent_sdk import JulesClient

# Your API key - set via environment variable or replace with your actual API key
# Example: export JULES_API_KEY="your-api-key-here"
API_KEY = os.environ.get("JULES_API_KEY", "your-api-key-here")

# Initialize the client with your API key
client = JulesClient(api_key=API_KEY)

try:
    # List available sources
    print("=== Available Sources ===")
    sources_result = client.sources.list(page_size=5)
    for source in sources_result["sources"]:
        print(f"Source ID: {source.id}")
        if source.github_repo:
            print(f"  GitHub: {source.github_repo.owner}/{source.github_repo.repo}")

    # Select a source (replace with your actual source ID)
    source_id = "sources/your-source-id"

    # Create a new session
    print("\n=== Creating Session ===")
    session = client.sessions.create(
        prompt="Fix the authentication bug in the login module",
        source=source_id,
        starting_branch="main",
        title="Authentication Bug Fix",
    )

    print(f"Session created: {session.id}")
    print(f"State: {session.state}")
    print(f"URL: {session.url}")

    # Monitor session progress
    print("\n=== Monitoring Progress ===")
    completed_session = client.sessions.wait_for_completion(
        session_id=session.id, poll_interval=5, timeout=600  # 10 minutes
    )

    print(f"Final state: {completed_session.state}")

    # Get session activities
    print("\n=== Session Activities ===")
    activities = client.activities.list_all(session.id)
    for activity in activities[:10]:  # Show first 10 activities
        print(f"\n{activity.description}")
        print(f"  Originator: {activity.originator}")
        print(f"  Time: {activity.create_time}")

        # Show activity-specific data
        if activity.agent_messaged:
            print(f"  Message: {activity.agent_messaged.get('agentMessage', '')[:100]}")
        elif activity.plan_generated:
            plan = activity.plan_generated.get("plan", {})
            steps = plan.get("steps", [])
            print(f"  Plan with {len(steps)} steps generated")

    # List all sessions
    print("\n=== Recent Sessions ===")
    sessions_result = client.sessions.list(page_size=5)
    for s in sessions_result["sessions"]:
        print(f"Session {s.id}: {s.state} - {s.title or s.prompt[:50]}")

finally:
    # Always close the client
    client.close()

File: jules-agent-sdk-go/examples/basic_usage.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"

	"github.com/sashimikun/jules-agent-sdk-go/jules"
)

func main() {
	// Get API key from environment variable
	apiKey := os.Getenv("JULES_API_KEY")
	if apiKey == "" {
		log.Fatal("JULES_API_KEY environment variable is required")
	}

	// Create a new Jules client
	client, err := jules.NewClient(apiKey)
	if err != nil {
		log.Fatalf("Failed to create client: %v", err)
	}
	defer client.Close()

	ctx := context.Background()

	// Example 1: List all sources
	fmt.Println("=== Listing Sources ===")
	sources, err := client.Sources.ListAll(ctx, "")
	if err != nil {
		log.Fatalf("Failed to list sources: %v", err)
	}

	for _, source := range sources {
		fmt.Printf("Source: %s (ID: %s)\n", source.Name, source.ID)
		if source.GitHubRepo != nil {
			fmt.Printf("  GitHub: %s/%s\n", source.GitHubRepo.Owner, source.GitHubRepo.Repo)
		}
	}

	// Example 2: Create a new session
	fmt.Println("\n=== Creating Session ===")
	session, err := client.Sessions.Create(ctx, &jules.CreateSessionRequest{
		Prompt: "Add a new feature to improve error handling",
		Source: sources[0].Name, // Use the first source
		Title:  "Improve Error Handling",
	})
	if err != nil {
		log.Fatalf("Failed to create session: %v", err)
	}

	fmt.Printf("Session created: %s\n", session.ID)
	fmt.Printf("State: %s\n", session.State)
	fmt.Printf("URL: %s\n", session.URL)

	// Example 3: Wait for session completion
	fmt.Println("\n=== Waiting for Completion ===")
	completedSession, err := client.Sessions.WaitForCompletion(ctx, session.ID, nil)
	if err != nil {
		log.Fatalf("Failed to wait for completion: %v", err)
	}

	fmt.Printf("Session completed: %s\n", completedSession.State)
	if completedSession.Output != nil && completedSession.Output.PullRequest != nil {
		fmt.Printf("Pull Request: %s\n", completedSession.Output.PullRequest.URL)
	}

	// Example 4: List activities for the session
	fmt.Println("\n=== Listing Activities ===")
	activities, err := client.Activities.ListAll(ctx, session.ID)
	if err != nil {
		log.Fatalf("Failed to list activities: %v", err)
	}

	for _, activity := range activities {
		fmt.Printf("Activity: %s\n", activity.Description)
		fmt.Printf("  Originator: %s\n", activity.Originator)
		if activity.CreateTime != nil {
			fmt.Printf("  Created: %s\n", activity.CreateTime)
		}
	}

	// Display client statistics
	fmt.Println("\n=== Client Statistics ===")
	stats := client.Stats()
	fmt.Printf("Total Requests: %d\n", stats["request_count"])
	fmt.Printf("Total Errors: %d\n", stats["error_count"])
}
